File: packages/plugin-agent/src/server/tools/sessions.ts
import type { AgentToolDefinition, AgentToolHandler, AgentToolContext, AgentToolResult } from '@purplesector/plugin-api';

export const sessionToolDefinitions: AgentToolDefinition[] = [
  {
    name: 'listSessions',
    description: 'List all sessions, optionally filtered by event ID.',
    inputSchema: {
      type: 'object',
      properties: {
        eventId: { type: 'string', description: 'Optional event ID to filter by' },
      },
    },
    category: 'sessions',
    mutating: false,
  },
  {
    name: 'getSession',
    description: 'Get details of a specific session by ID, including lap count.',
    inputSchema: {
      type: 'object',
      properties: {
        sessionId: { type: 'string', description: 'The session ID' },
      },
      required: ['sessionId'],
    },
    category: 'sessions',
    mutating: false,
  },
  {
    name: 'createSession',
    description: 'Create a new session within an event.',
    inputSchema: {
      type: 'object',
      properties: {
        eventId: { type: 'string', description: 'The event ID to create the session in' },
        name: { type: 'string', description: 'Session name (e.g. "Practice 1", "Qualifying")' },
        source: { type: 'string', description: 'Data source: "live" or "demo"' },
        vehicleId: { type: 'string', description: 'Optional vehicle ID' },
        vehicleConfigurationId: { type: 'string', description: 'Optional vehicle configuration ID' },
        vehicleSetupId: { type: 'string', description: 'Optional vehicle setup ID' },
      },
      required: ['eventId', 'name', 'source'],
    },
    category: 'sessions',
    mutating: true,
  },
  {
    name: 'updateSession',
    description: 'Update an existing session (name, status, vehicle assignment, etc.).',
    inputSchema: {
      type: 'object',
      properties: {
        sessionId: { type: 'string', description: 'The session ID to update' },
        name: { type: 'string' },
        status: { type: 'string', description: '"active" | "paused" | "archived"' },
        vehicleId: { type: 'string' },
        vehicleConfigurationId: { type: 'string' },
        vehicleSetupId: { type: 'string' },
      },
      required: ['sessionId'],
    },
    category: 'sessions',
    mutating: true,
  },
  {
    name: 'deleteSession',
    description: 'Delete a session and all its laps.',
    inputSchema: {
      type: 'object',
      properties: {
        sessionId: { type: 'string', description: 'The session ID to delete' },
      },
      required: ['sessionId'],
    },
    category: 'sessions',
    mutating: true,
  },
];

export function createSessionToolHandlers(prisma: any): Record<string, AgentToolHandler> {
  return {
    async listSessions(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const where: Record<string, unknown> = { userId: ctx.userId };
      if (args.eventId) where.eventId = args.eventId;

      const sessions = await prisma.session.findMany({
        where,
        include: { _count: { select: { laps: true } } },
        orderBy: { createdAt: 'desc' },
      });
      return { success: true, data: sessions, message: `Found ${sessions.length} sessions.` };
    },

    async getSession(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const session = await prisma.session.findFirst({
        where: { id: args.sessionId as string, userId: ctx.userId },
        include: {
          _count: { select: { laps: true } },
          event: { select: { id: true, name: true } },
          vehicle: { select: { id: true, name: true } },
          vehicleConfiguration: { select: { id: true, name: true } },
          vehicleSetup: { select: { id: true, name: true } },
        },
      });
      if (!session) return { success: false, message: 'Session not found.' };
      return { success: true, data: session };
    },

    async createSession(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const event = await prisma.event.findFirst({ where: { id: args.eventId as string, userId: ctx.userId } });
      if (!event) return { success: false, message: 'Event not found.' };

      const session = await prisma.session.create({
        data: {
          userId: ctx.userId,
          eventId: args.eventId as string,
          name: args.name as string,
          source: args.source as string,
          status: 'active',
          started: true,
          vehicleId: (args.vehicleId as string) || null,
          vehicleConfigurationId: (args.vehicleConfigurationId as string) || null,
          vehicleSetupId: (args.vehicleSetupId as string) || null,
        },
      });
      return { success: true, data: session, message: `Created session "${session.name}".` };
    },

    async updateSession(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const { sessionId, ...updates } = args;
      const existing = await prisma.session.findFirst({ where: { id: sessionId as string, userId: ctx.userId } });
      if (!existing) return { success: false, message: 'Session not found.' };

      const data: Record<string, unknown> = {};
      if (updates.name !== undefined) data.name = updates.name;
      if (updates.status !== undefined) data.status = updates.status;
      if (updates.vehicleId !== undefined) data.vehicleId = updates.vehicleId || null;
      if (updates.vehicleConfigurationId !== undefined) data.vehicleConfigurationId = updates.vehicleConfigurationId || null;
      if (updates.vehicleSetupId !== undefined) data.vehicleSetupId = updates.vehicleSetupId || null;

      const session = await prisma.session.update({ where: { id: sessionId as string }, data });
      return { success: true, data: session, message: `Updated session "${session.name}".` };
    },

    async deleteSession(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const existing = await prisma.session.findFirst({ where: { id: args.sessionId as string, userId: ctx.userId } });
      if (!existing) return { success: false, message: 'Session not found.' };
      await prisma.session.delete({ where: { id: args.sessionId as string } });
      return { success: true, message: `Deleted session "${existing.name}".` };
    },
  };
}

File: apps/web/apps/agent/tools/sessions.py
"""Agent tools for Session CRUD."""

from __future__ import annotations

from apps.events.models import Event, Session

from . import ToolDefinition, ToolContext, ToolResult, register_tools

definitions = [
    ToolDefinition(
        name="listSessions",
        description="List all sessions, optionally filtered by event ID.",
        input_schema={
            "type": "object",
            "properties": {
                "eventId": {"type": "string", "description": "Optional event ID to filter by"},
            },
        },
        category="sessions",
    ),
    ToolDefinition(
        name="getSession",
        description="Get details of a specific session by ID, including lap count.",
        input_schema={
            "type": "object",
            "properties": {
                "sessionId": {"type": "string", "description": "The session ID"},
            },
            "required": ["sessionId"],
        },
        category="sessions",
    ),
    ToolDefinition(
        name="createSession",
        description="Create a new session within an event.",
        input_schema={
            "type": "object",
            "properties": {
                "eventId": {"type": "string", "description": "The event ID to create the session in"},
                "name": {"type": "string", "description": 'Session name (e.g. "Practice 1", "Qualifying")'},
                "source": {"type": "string", "description": 'Data source: "live" or "demo"'},
                "vehicleId": {"type": "string", "description": "Optional vehicle ID"},
                "vehicleConfigurationId": {"type": "string", "description": "Optional vehicle configuration ID"},
                "vehicleSetupId": {"type": "string", "description": "Optional vehicle setup ID"},
            },
            "required": ["eventId", "name", "source"],
        },
        category="sessions",
        mutating=True,
    ),
    ToolDefinition(
        name="updateSession",
        description="Update an existing session (name, status, vehicle assignment, etc.).",
        input_schema={
            "type": "object",
            "properties": {
                "sessionId": {"type": "string", "description": "The session ID to update"},
                "name": {"type": "string"},
                "status": {"type": "string", "description": '"active" | "paused" | "archived"'},
                "vehicleId": {"type": "string"},
                "vehicleConfigurationId": {"type": "string"},
                "vehicleSetupId": {"type": "string"},
            },
            "required": ["sessionId"],
        },
        category="sessions",
        mutating=True,
    ),
    ToolDefinition(
        name="deleteSession",
        description="Delete a session and all its laps.",
        input_schema={
            "type": "object",
            "properties": {
                "sessionId": {"type": "string", "description": "The session ID to delete"},
            },
            "required": ["sessionId"],
        },
        category="sessions",
        mutating=True,
    ),
]


def _serialize_session(s: Session, lap_count: int | None = None) -> dict:
    d = {
        "id": str(s.id),
        "name": s.name,
        "source": s.source,
        "status": s.status,
        "started": s.started,
        "eventId": str(s.event_id),
        "vehicleId": str(s.vehicle_id) if s.vehicle_id else None,
        "vehicleConfigurationId": str(s.vehicle_configuration_id) if s.vehicle_configuration_id else None,
        "vehicleSetupId": str(s.vehicle_setup_id) if s.vehicle_setup_id else None,
        "createdAt": s.created_at.isoformat(),
    }
    if lap_count is not None:
        d["_count"] = {"laps": lap_count}
    return d


async def list_sessions(args: dict, ctx: ToolContext) -> ToolResult:
    from django.db.models import Count

    qs = Session.objects.filter(user_id=ctx.user_id).annotate(lap_count=Count("laps"))
    if args.get("eventId"):
        qs = qs.filter(event_id=args["eventId"])
    sessions = qs.order_by("-created_at")
    data = [_serialize_session(s, s.lap_count) for s in sessions]
    return ToolResult(success=True, data=data, message=f"Found {len(data)} sessions.")


async def get_session(args: dict, ctx: ToolContext) -> ToolResult:
    from django.db.models import Count

    try:
        session = (
            Session.objects.filter(id=args["sessionId"], user_id=ctx.user_id)
            .annotate(lap_count=Count("laps"))
            .select_related("event", "vehicle", "vehicle_configuration", "vehicle_setup")
            .get()
        )
    except Session.DoesNotExist:
        return ToolResult(success=False, message="Session not found.")
    d = _serialize_session(session, session.lap_count)
    d["event"] = {"id": str(session.event.id), "name": session.event.name}
    if session.vehicle:
        d["vehicle"] = {"id": str(session.vehicle.id), "name": session.vehicle.name}
    if session.vehicle_configuration:
        d["vehicleConfiguration"] = {"id": str(session.vehicle_configuration.id), "name": session.vehicle_configuration.name}
    if session.vehicle_setup:
        d["vehicleSetup"] = {"id": str(session.vehicle_setup.id), "name": session.vehicle_setup.name}
    return ToolResult(success=True, data=d)


async def create_session(args: dict, ctx: ToolContext) -> ToolResult:
    if not Event.objects.filter(id=args["eventId"], user_id=ctx.user_id).exists():
        return ToolResult(success=False, message="Event not found.")
    session = Session.objects.create(
        user_id=ctx.user_id,
        event_id=args["eventId"],
        name=args["name"],
        source=args["source"],
        status="active",
        started=True,
        vehicle_id=args.get("vehicleId") or None,
        vehicle_configuration_id=args.get("vehicleConfigurationId") or None,
        vehicle_setup_id=args.get("vehicleSetupId") or None,
    )
    return ToolResult(success=True, data=_serialize_session(session), message=f'Created session "{session.name}".')


async def update_session(args: dict, ctx: ToolContext) -> ToolResult:
    try:
        session = Session.objects.get(id=args["sessionId"], user_id=ctx.user_id)
    except Session.DoesNotExist:
        return ToolResult(success=False, message="Session not found.")

    for field in ("name", "status"):
        if field in args:
            setattr(session, field, args[field])
    for fk_field, attr in [("vehicleId", "vehicle_id"), ("vehicleConfigurationId", "vehicle_configuration_id"), ("vehicleSetupId", "vehicle_setup_id")]:
        if fk_field in args:
            setattr(session, attr, args[fk_field] or None)
    session.save()
    return ToolResult(success=True, data=_serialize_session(session), message=f'Updated session "{session.name}".')


async def delete_session(args: dict, ctx: ToolContext) -> ToolResult:
    try:
        session = Session.objects.get(id=args["sessionId"], user_id=ctx.user_id)
    except Session.DoesNotExist:
        return ToolResult(success=False, message="Session not found.")
    name = session.name
    session.delete()
    return ToolResult(success=True, message=f'Deleted session "{name}".')


handlers = {
    "listSessions": list_sessions,
    "getSession": get_session,
    "createSession": create_session,
    "updateSession": update_session,
    "deleteSession": delete_session,
}

register_tools(definitions, handlers)
